paddle/fluid/operators/fsp_op.h

```diff
@@ -0,0 +1,119 @@
+/* Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+#include "paddle/fluid/framework/op_registry.h"
+#include "paddle/fluid/operators/math/blas.h"
+#include "paddle/fluid/operators/math/math_function.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+
+template <typename DeviceContext, typename T>
+class FSPOpKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& context) const override {
+    auto* x = context.Input<Tensor>("X");
+    auto* y = context.Input<Tensor>("Y");
+    auto* output = context.Output<Tensor>("Out");
+    output->mutable_data<T>(context.GetPlace());
+    auto x_dims = x->dims();
+    auto y_dims = y->dims();
+
+    auto batch_size = x_dims[0];
+    auto x_channel = x_dims[1];
+    auto y_channel = y_dims[1];
+    auto height = x_dims[2];
+    auto width = x_dims[3];
+
+    auto blas = math::GetBlas<DeviceContext, T>(context);
+
+    for (int64_t batch = 0; batch < batch_size; ++batch) {
+      auto x_mat =
+          x->Slice(batch, batch + 1).Resize({x_channel, height * width});
+      auto y_mat =
+          y->Slice(batch, batch + 1).Resize({y_channel, height * width});
+      auto out_mat =
+          output->Slice(batch, batch + 1).Resize({x_channel, y_channel});
+
+      blas.MatMul(x_mat, false, y_mat, true,
+                  static_cast<T>(1.0 / (height * width)), &out_mat,
+                  static_cast<T>(0.0));
+    }
```

Comment:
Maybe can optimized by batch_mat_mul in next PR instead of `for { gemm }`

Reply:
Thx. Done.